lib/backend/k8s/syncer.go

```diff
@@ -0,0 +1,536 @@
+	poWatch, err := syn.kc.clientSet.Pods("").Watch(opts)
+	if err != nil {
+		panic(err)
+	}
+	opts = k8sapi.ListOptions{ResourceVersion: initialVersions.networkPolicyVersion}
+	npWatch, err := syn.kc.clientSet.NetworkPolicies("").Watch(opts)
+	if err != nil {
+		panic(err)
+	}
+
+	// Keep track of the latest resource versions.
+	latestVersions := initialVersions
+
+	nsChan := nsWatch.ResultChan()
+	poChan := poWatch.ResultChan()
+	npChan := npWatch.ResultChan()
+	var event watch.Event
+	var upd *model.Update
+	needsResync := false
+
+	log.Info("Starting Kubernetes API watch loop")
+	for {
+		// If we need to resync, do so.
+		if needsResync {
+			log.Warnf("Resync required - sending latest versions: %+v", latestVersions)
+			resyncChan <- latestVersions
+
+			// Wait to be told the new versions to watch.
+			log.Warn("Waiting for snapshot to complete")
+			latestVersions = <-initialVersionSource
+			log.Warnf("Snapshot complete after resync - start watch from %+v", latestVersions)
+
+			// Re-create the watches.
+			opts = k8sapi.ListOptions{ResourceVersion: initialVersions.namespaceVersion}
+			nsWatch, err = syn.kc.clientSet.Namespaces().Watch(opts)
+			if err != nil {
+				log.Warn("Failed to connect to API, retrying")
+				time.Sleep(1 * time.Second)
+				continue
+			}
+			opts = k8sapi.ListOptions{ResourceVersion: initialVersions.podVersion}
+			poWatch, err = syn.kc.clientSet.Pods("").Watch(opts)
+			if err != nil {
+				log.Warn("Failed to connect to API, retrying")
+				time.Sleep(1 * time.Second)
+				continue
+			}
+			opts = k8sapi.ListOptions{ResourceVersion: initialVersions.networkPolicyVersion}
+			npWatch, err = syn.kc.clientSet.NetworkPolicies("").Watch(opts)
+			if err != nil {
+				log.Warn("Failed to connect to API, retrying")
+				time.Sleep(1 * time.Second)
+				continue
+			}
+
+			nsChan = nsWatch.ResultChan()
+			poChan = poWatch.ResultChan()
+			npChan = npWatch.ResultChan()
+
+			// Successfull resync - reset the flag.
+			needsResync = false
+		}
+
+		// Select on the various watch channels.
+		select {
+		case event = <-nsChan:
+			log.Debugf("Incoming Namespace watch event. Type=%s", event.Type)
+			if needsResync = syn.eventTriggersResync(event); needsResync {
+				// We need to resync.  Break out of the inner for loop, back
+				// into the sync loop.
+				log.Warn("Event triggered resync: %+v", event)
+				break
+			}
+
+			// Event is OK - parse it.
+			upds := syn.parseNamespaceEvent(event)
+			for _, k := range upds {
+				if k == nil {
+					// This can return nil when the event is not
+					// one we care about.
+					continue
+				}
+				updateChan <- k
+				latestVersions.namespaceVersion = k.KVPair.Revision.(string)
+			}
+			continue
+		case event = <-poChan:
+			log.Debugf("Incoming Pod watch event. Type=%s", event.Type)
+			if needsResync = syn.eventTriggersResync(event); needsResync {
+				// We need to resync.  Break out of the inner for loop, back
```

Comment:
I tink this actually breaks out of the select, which confused me the first time I used break within a select!